File: game/scripts/vscripts/abilities/heroes/puck/puck_basic_attack/puck_basic_attack.lua
puck_basic_attack = class({})
puck_ex_basic_attack = class({})
puck_basic_attack_related = class({})

LinkLuaModifier("modifier_puck_basic_attack_cooldown", "abilities/heroes/puck/puck_basic_attack/modifier_puck_basic_attack_cooldown", LUA_MODIFIER_MOTION_NONE)
LinkLuaModifier("modifier_puck_basic_attack_buff", "abilities/heroes/puck/puck_basic_attack/modifier_puck_basic_attack_cooldown", LUA_MODIFIER_MOTION_NONE)

LinkLuaModifier("modifier_puck_fairy_dust", "abilities/heroes/puck/modifier_puck_fairy_dust", LUA_MODIFIER_MOTION_NONE)
LinkLuaModifier("modifier_puck_ex_basic_attack", "abilities/heroes/puck/puck_basic_attack/modifier_puck_ex_basic_attack", LUA_MODIFIER_MOTION_NONE)
LinkLuaModifier("modifier_puck_basic_attack_thinker", "abilities/heroes/puck/puck_basic_attack/modifier_puck_basic_attack_thinker", LUA_MODIFIER_MOTION_NONE)
LinkLuaModifier("modifier_puck_ex_basic_attack_shield", "abilities/heroes/puck/puck_basic_attack/modifier_puck_ex_basic_attack_shield", LUA_MODIFIER_MOTION_NONE)



function puck_basic_attack:GetCastPoint()
	if IsServer() then
		return self.BaseClass.GetCastPoint(self) + self:GetCaster():GetAttackAnimationPoint()
	end
end

function puck_basic_attack:GetCooldown(iLevel)
	if IsServer() then
        local attacks_per_second = self:GetCaster():GetAttacksPerSecond()
        local attack_speed = (1 / attacks_per_second)
		
		return self.BaseClass.GetCooldown(self, self:GetLevel()) + attack_speed
	end
end

function puck_basic_attack:GetCastAnimationCustom()	    return ACT_DOTA_ATTACK end
function puck_basic_attack:GetPlaybackRateOverride() 	return 1.0 end
function puck_basic_attack:GetCastPointSpeed() 		    return 10 end

function puck_basic_attack:GetIntrinsicModifierName()
	return "modifier_puck_basic_attack_cooldown"
end

function puck_basic_attack:OnSpellStart()
	self:LaunchProjectile(self:GetCaster():GetOrigin(), CustomAbilitiesLegacy:GetCursorPosition(self))
end

function puck_basic_attack:LaunchProjectile(origin, point)
	local caster = self:GetCaster()

	local projectile_speed = self:GetSpecialValueFor("projectile_speed")
	local projectile_direction = Direction2D(origin, point)

    local mana_gain_pct = self:GetSpecialValueFor("mana_gain_pct")
    local fairy_dust_duration = self:GetSpecialValueFor("fairy_dust_duration")
    local fairy_dust_slow_pct = self:GetSpecialValueFor("fairy_dust_slow_pct")

	local is_charged = caster:FindModifierByName("modifier_puck_basic_attack_cooldown"):IsCooldownReady()
	
	local projectile_particle = "particles/puck/puck_base_attack.vpcf"

	if is_charged then
		projectile_particle = "particles/puck/puck_base_attack_alternative.vpcf"
	end

	CustomEntitiesLegacy:ProjectileAttack(caster, {
		bIsBasicAttack = true,
		tProjectile = {
			EffectName = projectile_particle,
			vSpawnOrigin = origin + Vector(projectile_direction.x * 45, projectile_direction.y * 45, 96),
			fDistance = self:GetSpecialValueFor("projectile_distance") ~= 0 and self:GetSpecialValueFor("projectile_distance") or self:GetCastRange(Vector(0,0,0), nil),
			fStartRadius = self:GetSpecialValueFor("hitbox"),
			Source = caster,
			vVelocity = projectile_direction * projectile_speed,
			UnitBehavior = PROJECTILES_DESTROY,
			TreeBehavior = PROJECTILES_NOTHING,
			WallBehavior = PROJECTILES_DESTROY,
			GroundBehavior = PROJECTILES_NOTHING,
			fGroundOffset = 0,
			UnitTest = function(_self, unit) return unit:GetUnitName() ~= "npc_dummy_unit" and not CustomEntitiesLegacy:Allies(_self.Source, unit) end,
			OnUnitHit = function(_self, unit)
				CustomEntitiesLegacy:AttackWithBaseDamage(caster, {
					hTarget = unit,
					hAbility = self,
				})
				
				if is_charged then
					unit:AddNewModifier(_self.Source, nil, "modifier_puck_fairy_dust", { duration = fairy_dust_duration, slow_pct = fairy_dust_slow_pct })
				end

				if _self.Source == caster and not CustomEntitiesLegacy:IsObstacle(unit) then
					if CustomEntitiesLegacy:ProvidesMana(unit) then
						CustomEntitiesLegacy:GiveManaAndEnergyPercent(caster, mana_gain_pct, true)
					end
					if self:GetLevel() >=2 then
						caster:FindModifierByName("modifier_puck_basic_attack_cooldown"):Replenish()
					end
				end
			end,
			OnFinish = function(_self, pos)
				self:PlayEffectsOnFinish(pos, is_charged)
			end,
		}
	})

	self:PlayEffectsOnCast(is_charged)
end

function puck_basic_attack:PlayEffectsOnCast(isCharged)
	EmitSoundOn("Hero_Puck.Attack", self:GetCaster())
	if isCharged then
		EmitSoundOn("Hero_Oracle.FortunesEnd.Attack", self:GetCaster())
	end
end

function puck_basic_attack:PlayEffectsOnFinish(pos, is_charged)
	local caster = self:GetCaster()

	EmitSoundOnLocationWithCaster(pos, "Hero_Puck.ProjectileImpact", caster)

	local particle_cast = "particles/puck/puck_base_attack_explosion.vpcf"
	if is_charged then
		particle_cast = "particles/puck/puck_base_attack_alternative_explosion.vpcf"
	end
	
	local effect_cast = ParticleManager:CreateParticle(particle_cast, PATTACH_ABSORIGIN, caster)
	ParticleManager:SetParticleControl(effect_cast, 3, pos)
	ParticleManager:ReleaseParticleIndex(effect_cast)
end

function puck_basic_attack:OnUpgrade()
	local related = self:GetCaster():FindAbilityByName('puck_basic_attack_related')
	
	if self:GetLevel() > related:GetLevel() then
		related:UpgradeAbility(true)
	end
end

puck_basic_attack_related.GetCooldown = puck_basic_attack.GetCooldown
puck_basic_attack_related.PlayEffectsOnCast = puck_basic_attack.PlayEffectsOnCast

function puck_basic_attack_related:GetCastAnimationCustom()	    return ACT_DOTA_CAST_ABILITY_2 end
function puck_basic_attack_related:GetPlaybackRateOverride() 	return 1.0 end
function puck_basic_attack_related:GetCastPointSpeed() 		    return 10 end

function puck_basic_attack_related:OnSpellStart()
	local caster = self:GetCaster()
	local origin = caster:GetAbsOrigin()
	local point = ClampPosition(origin, CustomAbilitiesLegacy:GetCursorPosition(self), self:GetCastRange(Vector(0,0,0), nil), nil)
	local charged = caster:FindModifierByName("modifier_puck_basic_attack_cooldown"):IsCooldownReady() and 1 or 0
	
	CreateModifierThinker(
		caster, --hCaster
		self, --hAbility
		"modifier_puck_basic_attack_thinker", --modifierName
		{
			charged = charged
		},
		point, --vOrigin
		caster:GetTeamNumber(), --nTeamNumber
		false --bPhantomBlocker
	)
	
    EFX('particles/econ/items/invoker/invoker_ti7/invoker_ti7_alacrity_cast.vpcf', PATTACH_CUSTOMORIGIN, caster, {
		cp0 = {
			ent = caster,
			point = 'attach_hitloc'
		},
		cp1 = point + Vector(0, 0, 1000),
		cp2 = point + Vector(0, 0, 1000),
        release = true,
	})
	self:PlayEffectsOnCast(charged == 1)
end

function puck_basic_attack_related:OnUpgrade()
	local related = self:GetCaster():FindAbilityByName('puck_basic_attack')
	
	if self:GetLevel() > related:GetLevel() then
		related:UpgradeAbility(true)
	end
end

function puck_ex_basic_attack:OnSpellStart()
	local caster = self:GetCaster()
	local duration = self:GetSpecialValueFor("duration")
	caster:AddNewModifier(caster, self, "modifier_puck_ex_basic_attack", { duration = duration })

	EFX("particles/puck/puck_ex_base_attack.vpcf", PATTACH_ABSORIGIN_FOLLOW, caster, {
		release = true
	})
	
    EmitSoundOn("puck_puck_laugh_01", caster)
	EmitSoundOn('Hero_Puck.EtherealJaunt', caster)
	EmitSoundOn("Hero_Puck.Phase_Shift", caster)
end


if IsClient() then require("wrappers/abilities") end
Abilities.Castpoint(puck_basic_attack)
Abilities.Castpoint(puck_basic_attack_related)

File: src/vscripts/abilities/heroes/puck/puck_basic_attack.ts
import { registerAbility, registerModifier } from "../../../lib/dota_ts_adapter";
import { ModifierCombatEvents } from "../../../modifiers/modifier_combat_events";
import { ModifierCooldown } from "../../../modifiers/modifier_cooldown";
import { ModifierShield } from "../../../modifiers/modifier_shield";
import { ModifierThinker, ModifierThinkerParams } from "../../../modifiers/modifier_thinker";
import { attackWithBaseDamage, clampPosition, direction2D, giveManaAndEnergyPercent, isGem, isObstacle, replenishEFX } from "../../../util";
import { CustomAbility } from "../../framework/custom_ability";
import { CustomModifier } from "../../framework/custom_modifier";

class PuckBasicAttackCommon extends CustomAbility {
    GetCooldown(level: number) {
        if (IsServer()) {
            const attacksPerSecond = this.caster.GetAttacksPerSecond();
            const attackSpeed = 1 / attacksPerSecond;
            return super.GetCooldown(level) + attackSpeed;
        }

        return super.GetCooldown(level);
    }

    PlayEffectsOnCast(charged: boolean) {
        EmitSoundOn("Hero_Puck.Attack", this.caster);
        if (charged) {
            EmitSoundOn("Hero_Oracle.FortunesEnd.Attack", this.caster);
        }
    }
}

@registerAbility("puck_basic_attack")
export class PuckBasicAttack extends PuckBasicAttackCommon {
    GetCastPoint() {
        if (IsServer()) {
            return super.GetCastPoint() + this.caster.GetAttackAnimationPoint();
        }

        return super.GetCastPoint();
    }

    GetAnimation() {
        return GameActivity.DOTA_ATTACK;
    }

    GetPlaybackRateOverride() {
        return 1.0;
    }

    GetCastingCrawl() {
        return 10;
    }

    GetIntrinsicModifierName() {
        return ModifierPuckBasicAttack.name;
    }

    OnSpellStart() {
        const origin = this.caster.GetAbsOrigin();
        const point = CustomAbilitiesLegacy.GetCursorPosition(this);

        this.LaunchProjectile(origin, point);
    }

    LaunchProjectile(origin: Vector, point: Vector) {
        const manaGainPct = this.GetSpecialValueFor("mana_gain_pct");
        const projectileSpeed = this.GetSpecialValueFor("projectile_speed");
        const projectileDirection = direction2D(origin, point);

        const fairyDustDuration = this.GetSpecialValueFor("fairy_dust_duration");
        const fairyDustSlowPct = this.GetSpecialValueFor("fairy_dust_slow_pct");

        const isCharged = ModifierPuckBasicAttack.findOne(this.caster)?.IsCooldownReady() ?? false;

        this.ProjectileAttack({
            source: this.caster,
            attackType: "basic",
            effectName: isCharged ? "particles/puck/puck_base_attack_alternative.vpcf" : "particles/puck/puck_base_attack.vpcf",
            spawnOrigin: origin.__add(Vector(projectileDirection.x * 45, projectileDirection.y * 45, 96)),
            velocity: projectileDirection.__mul(projectileSpeed),
            groundOffset: 0,
            unitTest: (unit, projectile) =>
                unit.GetUnitName() != "npc_dummy_unit" && !CustomEntitiesLegacy.Allies(projectile.getSource(), unit),
            onUnitHit: (unit, projectile) => {
                attackWithBaseDamage({
                    source: projectile.getSource(),
                    target: unit,
                    ability: this
                });

                if (isCharged) {
                    ModifierPuckFairyDust.apply(unit, projectile.getSource(), this, {
                        duration: fairyDustDuration,
                        slowPct: fairyDustSlowPct
                    });
                }

                if (projectile.getSource() == this.caster && !isObstacle(unit)) {
                    if (!isGem(unit)) {
                        giveManaAndEnergyPercent(projectile.getSource(), manaGainPct, true);
                    }

                    if (this.GetLevel() >= 2) {
                        ModifierPuckBasicAttack.findOne(this.caster)?.Replenish();
                    }
                }
            },
            onFinish: projectile => {
                this.PlayEffectsOnFinish(projectile.getPosition(), isCharged);
            }
        });
    }

    PlayEffectsOnFinish(position: Vector, charged: boolean) {
        EmitSoundOnLocationWithCaster(position, "Hero_Puck.ProjectileImpact", this.caster);
        const path = charged
            ? "particles/puck/puck_base_attack_alternative_explosion.vpcf"
            : "particles/puck/puck_base_attack_explosion.vpcf";
        const particleId = ParticleManager.CreateParticle(path, ParticleAttachment.ABSORIGIN, this.caster);
        ParticleManager.SetParticleControl(particleId, 3, position);
        ParticleManager.ReleaseParticleIndex(particleId);
    }

    // OnUpgrade(){
    // 	const related = this.caster.FindAbilityByName('puck_basic_attack_related')
    // 	if(this.GetLevel() > related:GetLevel()){
    // 		related:UpgradeAbility(true)
    // 	}
    // }
}

@registerAbility("puck_basic_attack_related")
export class PuckBasicAttackRelated extends PuckBasicAttackCommon {
    GetAnimation() {
        return GameActivity.DOTA_CAST_ABILITY_2;
    }

    GetPlaybackRateOverride() {
        return 1.0;
    }

    GetCastingCrawl() {
        return 10;
    }

    OnSpellStart() {
        const origin = this.caster.GetAbsOrigin();
        const cursor = CustomAbilitiesLegacy.GetCursorPosition(this);
        const point = clampPosition(origin, cursor, { maxRange: this.GetCastRange(Vector(0, 0, 0), undefined) });
        const puckExBasicAttack = PuckExBasicAttack.findOne(this.caster)?.GetSpecialValueFor("delay_time") ?? 0;

        ModifierPuckExBasicAttackThinker.createThinker(this.caster, this, point, {
            radius: this.GetSpecialValueFor("radius"),
            status: ModifierPuckBasicAttack.findOne(this.caster)?.IsCooldownReady() ? "charged" : "notCharged",
            delayTime: puckExBasicAttack
        });

        EFX("particles/econ/items/invoker/invoker_ti7/invoker_ti7_alacrity_cast.vpcf", ParticleAttachment.CUSTOMORIGIN, this.caster, {
            cp0: {
                ent: this.caster,
                point: "attach_hitloc"
            },
            cp1: point + Vector(0, 0, 1000),
            cp2: point + Vector(0, 0, 1000),
            release: true
        });
        // 	this.PlayEffectsOnCast(charged == 1)
    }

    // OnUpgrade(){
    // 	const related = this.caster.FindAbilityByName('puck_basic_attack')
    // 	if(this.GetLevel() > related:GetLevel()){
    // 		related:UpgradeAbility(true)
    // 	}
    // }
}

@registerModifier({ customNameForI18n: "modifier_puck_basic_attack_cooldown" })
export class ModifierPuckBasicAttack extends ModifierCooldown {
    DeclareFunctions() {
        return [ModifierFunction.PREATTACK_BONUS_DAMAGE];
    }

    OnBasicAttackStarted() {
        if (this.GetRemainingTime() > 0) {
            return;
        }

        if (IsServer()) {
            ModifierPuckBasicAttackBuff.apply(this.parent, this.parent, this.ability, {});
        }

        this.StartCooldown();
    }

    GetModifierPreAttack_BonusDamage() {
        if (!this.IsCooldownReady()) {
            return 0;
        }

        return this.Value("charged_damage");
    }

    OnReplenish() {
        if (IsServer()) {
            replenishEFX(this.parent);
        }
    }

    GetReplenishTime() {
        return this.Value("replenish_time");
    }
}
// if(IsClient()){ require("wrappers/modifiers") }
// Modifiers.Cooldown(modifier_puck_basic_attack_cooldown)
// Modifiers.OnEvent(modifier_puck_basic_attack_cooldown)

@registerModifier({ customNameForI18n: "modifier_puck_basic_attack_buff" })
class ModifierPuckBasicAttackBuff extends ModifierCombatEvents {
    IsHidden() {
        return true;
    }

    OnBasicAttackEnded() {
        this.Destroy();
    }

    // GetPreAttackDamage() {
    //     return this.Value("charged_damage");
    // }
}
// Modifiers.OnEvent(modifier_puck_basic_attack_buff)
// Modifiers.PreAttackDamage(modifier_puck_basic_attack_buff)

@registerModifier({ customNameForI18n: "modifier_puck_fairy_dust" })
export class ModifierPuckFairyDust extends CustomModifier {
    particleId?: ParticleID;

    OnCreated(params: { slowPct: number }) {
        if (IsServer()) {
            this.SetStackCount(params.slowPct);
            this.particleId = ParticleManager.CreateParticle(
                "particles/puck/puck_mobility_trail.vpcf",
                ParticleAttachment.ABSORIGIN_FOLLOW,
                this.parent
            );
        }
    }

    OnDestroy() {
        if (IsServer() && this.particleId) {
            ParticleManager.DestroyParticle(this.particleId, false);
            ParticleManager.ReleaseParticleIndex(this.particleId);
        }
    }

    OnRefresh(params: { slowPct: number }) {
        if (IsServer()) {
            if (params.slowPct > this.GetStackCount()) {
                this.SetStackCount(params.slowPct);
            }
        }
    }

    DeclareFunctions() {
        return [ModifierFunction.MOVESPEED_BONUS_PERCENTAGE];
    }

    GetModifierMoveSpeedBonus_Percentage() {
        return -this.GetStackCount();
    }

    GetStatusEffectName() {
        return "particles/status_fx/status_effect_combo_breaker.vpcf";
    }

    GetTexture() {
        return "modifier_puck_fairy_dust";
    }

    // function modifier_puck_fairy_dust:GetStatusLabel() return "Fairy Dust" }
    // function modifier_puck_fairy_dust:GetStatusPriority() return 4 }
    // function modifier_puck_fairy_dust:GetStatusStyle() return "FairyDust" }
}
// if IsClient() then require("wrappers/modifiers") }
// Modifiers.Status(modifier_puck_fairy_dust)

@registerAbility("puck_ex_basic_attack")
class PuckExBasicAttack extends CustomAbility {
    OnSpellStart() {
        const duration = this.GetSpecialValueFor("duration");
        ModifierPuckExBasicAttack.apply(this.caster, this.caster, this, { duration: duration });
        EFX("particles/puck/puck_ex_base_attack.vpcf", ParticleAttachment.ABSORIGIN_FOLLOW, this.caster, {
            release: true
        });
        EmitSoundOn("puck_puck_laugh_01", this.caster);
        EmitSoundOn("Hero_Puck.EtherealJaunt", this.caster);
        EmitSoundOn("Hero_Puck.Phase_Shift", this.caster);
    }
}

@registerModifier({ customNameForI18n: "modifier_puck_ex_basic_attack" })
class ModifierPuckExBasicAttack extends CustomModifier {
    particleIds: ParticleID[] = [];

    OnCreated() {
        if (IsServer()) {
            this.parent.SwapAbilities(PuckBasicAttack.name, PuckBasicAttackRelated.name, false, true);
            this.particleIds.push(
                ParticleManager.CreateParticle(
                    "particles/econ/items/mirana/mirana_ti8_immortal_mount/mirana_ti8_immortal_leap_trail_embers.vpcf",
                    ParticleAttachment.ABSORIGIN_FOLLOW,
                    this.parent
                )
            );
            this.particleIds.push(
                ParticleManager.CreateParticle("particles/puck/puck_mobility_trail.vpcf", ParticleAttachment.ABSORIGIN_FOLLOW, this.parent)
            );
        }
    }

    OnDestroy() {
        StopSoundOn("Hero_Puck.Phase_Shift", this.parent);
        if (IsServer()) {
            this.parent.SwapAbilities(PuckBasicAttack.name, PuckBasicAttackRelated.name, true, false);

            for (const particleId of this.particleIds) {
                ParticleManager.DestroyParticle(particleId, false);
                ParticleManager.ReleaseParticleIndex(particleId);
            }
            ParticleManager.ReleaseParticleIndex(
                ParticleManager.CreateParticle(
                    "particles/econ/items/mirana/mirana_ti8_immortal_mount/mirana_ti8_immortal_leap_start_embers.vpcf",
                    ParticleAttachment.ABSORIGIN_FOLLOW,
                    this.parent
                )
            );
            this.ability.StartCooldown(this.ability.GetCooldown(0));
        }
    }
    GetStatusEffectName() {
        return "particles/status_fx/status_effect_ghost.vpcf";
    }
    // function modifier_puck_ex_basic_attack.GetStatusLabel() return "Fairy Form" }
    // function modifier_puck_ex_basic_attack.GetStatusPriority() return 4 }
    // function modifier_puck_ex_basic_attack.GetStatusStyle() return "FairyDust" }
}
// if(IsClient()){ require("wrappers/modifiers") }
// Modifiers.Status(modifier_puck_ex_basic_attack)

//@Refactor maybe I should use the PuckExBasicAttack as the ability instead of the PuckBasicAttackRelated
@registerModifier({ customNameForI18n: "modifier_puck_basic_attack_thinker" })
class ModifierPuckExBasicAttackThinker extends ModifierThinker {
    charged!: boolean;
    origin!: Vector;
    duration!: number;
    shieldPerHit!: number;
    healPerHit!: number;
    fairyDustDuration!: number;
    fairyDustSlowPct!: number;
    manaGainPct!: number;

    exBasicAttack?: PuckExBasicAttack;

    OnCreated(params: ModifierThinkerParams & { status: "charged" | "notCharged" }) {
        super.OnCreated(params);

        if (IsServer()) {
            this.origin = this.parent.GetAbsOrigin();
            this.manaGainPct = this.ability.GetSpecialValueFor("mana_gain_pct");
            this.exBasicAttack = PuckExBasicAttack.findOne(this.caster);

            if (!this.exBasicAttack) {
                print("[ERROR] On PuckBasicAttackRelated: Can't find PuckExBasicAttack to get the values!");
                this.Destroy();
                return;
            }

            this.duration = this.exBasicAttack.GetSpecialValueFor("duration");
            this.shieldPerHit = this.exBasicAttack.GetSpecialValueFor("shield_per_hit");
            this.healPerHit = this.exBasicAttack.GetSpecialValueFor("heal_per_hit");
            this.fairyDustDuration = this.exBasicAttack.GetSpecialValueFor("fairy_dust_duration");
            this.fairyDustSlowPct = this.exBasicAttack.GetSpecialValueFor("fairy_dust_slow_pct");
            this.charged = params.status == "charged";
        }
    }

    OnReady() {
        EmitSoundOn("Hero_Leshrac.Lightning_Storm", this.parent);
        const color = this.charged ? Vector(100, 0, 0) : Vector(0, 0, 0);
        let giveMana = false;

        if (this.charged) {
            EFX(
                "particles/econ/generic/generic_aoe_explosion_sphere_1/generic_aoe_explosion_sphere_1.vpcf",
                ParticleAttachment.WORLDORIGIN,
                undefined,
                {
                    cp0: this.origin,
                    cp1: Vector(200, 0, 0),
                    cp2: Vector(2.0, 1, 1),
                    cp3: Vector(255, 0, 229),
                    cp4: Vector(250, 0, 255),
                    release: true
                }
            );
        }

        EFX("particles/puck/puck_basic_attack_related.vpcf", ParticleAttachment.WORLDORIGIN, undefined, {
            cp0: this.origin,
            cp1: this.origin,
            cp6: color,
            release: true
        });

        this.ability.AoeAttack({
            attackType: "basic",
            origin: this.origin,
            radius: this.radius,
            effect: target => {
                this.caster.PerformAttack(target, true, true, true, true, false, false, true);
                ModifierPuckFairyDust.apply(this.caster, this.caster, this.ability, {
                    duration: this.fairyDustDuration,
                    slowPct: this.fairyDustSlowPct
                });

                if (!isObstacle(target)) {
                    giveMana = true;
                }
            }
        });

        if (giveMana) {
            if (this.ability.GetLevel() >= 2) {
                ModifierPuckBasicAttack.findOne(this.caster)?.Replenish();
            }

            if (this.exBasicAttack && this.exBasicAttack.GetLevel() >= 2) {
                this.caster.Heal(this.healPerHit, this.ability);
            }

            ModifierPuckExBasicAttackShield.apply(this.caster, this.caster, undefined, {
                duration: 6.0,
                damageBlock: this.shieldPerHit
            });

            giveManaAndEnergyPercent(this.caster, this.manaGainPct, true);
        }
        this.Destroy();
    }
}

@registerModifier({ customNameForI18n: "modifier_puck_ex_basic_attack_shield" })
class ModifierPuckExBasicAttackShield extends ModifierShield {
    // function modifier_puck_ex_basic_attack_shield:GetStatusLabel() return "Fairy Shield" }
    // function modifier_puck_ex_basic_attack_shield:GetStatusPriority() return 3 }
    // function modifier_puck_ex_basic_attack_shield:GetStatusStyle() return "Shield" }
    // function modifier_puck_ex_basic_attack_shield:GetStackeable() return true }
}
// if(IsClient()){ require("wrappers/modifiers") }
// Modifiers.Status(modifier_puck_ex_basic_attack_shield)
// Modifiers.Shield(modifier_puck_ex_basic_attack_shield)
